bkbit/scripts/gfftranslator.py

```diff
@@ -4,133 +4,346 @@
+                                                    authority = authority_type
+                                                )
+    
+    # GENERATE GENE ANNOTATION OBJECT #
+    if authority_type.value == 'NCBI':
+        curr_gene_annots = generate_ncbi_gene_annotation(df_gene, curr_genome_annot, curr_org_taxon, taxon_local_unique_identifier)
+    else: 
+        curr_gene_annots = generate_ensembl_gene_annotation(df_gene, curr_genome_annot, curr_org_taxon, taxon_local_unique_identifier)
+    print(f'NUMBER OF GENE ANNOTATIONS: {len(curr_gene_annots)}')
+
+    # COMBINE ALL GENERATED OBJECTS RELATED TO THIS GFF FILE #
+    fileObjects.append(genomeAssemblyObjects[assembly_local_unique_identifier])
+    fileObjects.append(curr_genome_annot)
+    fileObjects.append(curr_org_taxon)
+    fileObjects.extend(list(digests.values())) 
+    fileObjects.extend(curr_gene_annots)
+    output_ser = os.path.join(output_dir, gene_name +'.json')
+    serialize_annotation_collection(fileObjects, output_ser)
+    print(f'Serialized GeneAnnotation objects saved here: {output_ser}')
+
+
+def generate_ensembl_gene_annotation(df, genomeAnnot, orgTaxon, taxon_id):
+    """Creates a gene annotation object for every row in the provided dataframe.
 
-    # authority specific mapping and transforms
-    if authority == 'NCBI' :
-        out_cols = ['genome_annotation_label','gene_identifier_prefix','GeneID','Name','description','gene_biotype', 'taxon_unique_identifier']
-    elif authority == 'Ensembl':
-        out_cols = ['genome_annotation_label','gene_identifier_prefix','gene_id','Name','description','biotype', 'taxon_unique_identifier']
+    Parameters:
+    df (pandas.DataFrame): dataframe containing gene annotations and respective metadata
+    genomeAnnot (GenomeAnnotation): reference to GenomeAnnotation object 
+    orgTaxon (OrganismTaxon): reference to OrganismTaxon object 
+    taxon_id (str): taxon id 
 
-    # ATTRIBUTES FOR GENE ANNOTATION OBJECT:
-    cols = ['genome_annotation_label','gene_identifier_prefix','gene_local_unique_identifier','name','description', 'gene_biotype', 'taxon_unique_identifier']
+    Returns:
+    gene_annotations (List[GeneAnnotations]): list of gene annotations created
 
-    # create output dataframe
-    out_df = df_gene[out_cols]
-    out_df.columns = cols
+    """
+    gene_annotations = []
 
-    fname = 'gene_annotation_%s-%s-%s.csv' % (authority, str(taxon_local_unique_id), str(version))
-    file = os.path.join(output_dir, fname)
-    out_df.to_csv(file, index=False)
+    for row in df.itertuples():
+        attributes = dict([x.split('=') for x in row.attributes.split(';')])
+        description = attributes.get("description")
+        if description is not None:
+            description = urllib.parse.unquote(description)
+            description = re.sub(r"\[source .*?\]", "", description)
```

Comment:
Have you checked this on actual GFF3 files? Lydia's documentaiton capitalized "Source" (but maybe that was a typo)

Reply:
typo on my end; "Source" is correct